lib/cli/src/commands/create_exe.rs

```diff
@@ -68,268 +61,187 @@ pub struct CreateExe {
-                        Artifact::generate_object(
-                            compiler, &data, prefixer, &target, tunables, features,
-                        )?;
-
-                    let header_file_src = crate::c_gen::staticlib_header::generate_header_file(
-                        &module_info,
-                        &*symbol_registry,
-                        metadata_length,
-                    );
-                    // Write object file with functions
-                    let object_file_path: std::path::PathBuf =
-                        std::path::Path::new("functions.o").into();
-                    let mut writer = BufWriter::new(File::create(&object_file_path)?);
-                    obj.write_stream(&mut writer)
-                        .map_err(|err| anyhow::anyhow!(err.to_string()))?;
-                    writer.flush()?;
-                    // Write down header file that includes pointer arrays and the deserialize function
-                    let mut writer = BufWriter::new(File::create("static_defs.h")?);
-                    writer.write_all(header_file_src.as_bytes())?;
-                    writer.flush()?;
-                    if let Some(setup) = cross_compilation.as_ref() {
-                        self.compile_zig(
-                            output_path,
-                            &[object_file_path],
-                            &[std::path::Path::new("static_defs.h").into()],
-                            setup,
-                            &[],
-                            None,
-                            None,
-                        )?;
-                    } else {
-                        self.link(
-                            output_path,
-                            object_file_path,
-                            std::path::Path::new("static_defs.h").into(),
-                            &[],
-                            None,
-                            None,
-                        )?;
-                    }
-                }
-            }
+            // wasm file
+            let atoms = prepare_directory_from_single_wasm_file(
+                &input_path,
+                &tempdir,
+                &self.compiler,
+                &cross_compilation.target,
+                &self.cpu_features,
+                object_format,
+                &self.precompiled_atom,
+                self.debug_dir.is_some(),
+            )?;
+            get_module_infos(&tempdir, &atoms, object_format)?;
+            let mut entrypoint = get_entrypoint(&tempdir)?;
+            create_header_files_in_dir(&tempdir, &mut entrypoint, &atoms, &self.precompiled_atom)?;
+            link_exe_from_dir(
+                &tempdir,
+                output_path,
+                &cross_compilation,
+                &self.libraries,
+                self.debug_dir.is_some(),
+                &atoms,
+                &self.precompiled_atom,
+            )?;
```

Comment:
> Or even matching on the extension so people don't get errors like "invalid wasm file" when actually their *.webc file failed to parse because it had a bad checksum or signature or whatever.

No, I wouldn't match on the extension, because pirita files in the future may not even have an extension (where the filename is just a hash for example).

Reply:
Implemented in https://github.com/wasmerio/wasmer/pull/3430/commits/c450fc28ea08beac935893cbf2088a578d403fe2